### ISSUE TITLE
k3s can't pull images from private registry but crictl is able to

### ISSUE THREAD

**Issue by MyIgel** [k3s-io/k3s#2787]
**Environmental Info:**
K3s Version: 
```
k3s version v1.20.0+k3s2 (2ea6b163)
go version go1.15.5
```

Node(s) CPU architecture, OS, and Version: 
`Linux hostname 4.19.0-12-amd64 #1 SMP Debian 4.19.152-1 (2020-10-18) x86_64 GNU/Linux`

Cluster Configuration:
1 master, 1 worker (both have the problem)

**Describe the bug:**
k3s can't pull a image (401 Unauthorized) from my gitlab but crictl can

**Steps To Reproduce:**
- Installed K3s
- Konfigured /etc/rancher/k3s/registries.yaml 
```yaml
configs:
  "my.registry.tld":
    auth:
      username: user
      password: token
```

**Expected behavior:**
If crictl can pull images the cluster should also be able to?

**Actual behavior:**
The k3s logs show
```
Jan 07 21:31:22 area51 k3s[15803]: E0107 21:31:22.850050   15803 pod_workers.go:191] Error syncing pod 3c8b5ed2-fc7c-47e0-95e8-30b56df50794 ("app-deployment-bfcf44fdd-jrbwv_project-production(3c8b5ed2-fc7c-47e0-95e8-30b56df50794)"), skipping: failed to "StartContainer" for "app" with ErrImagePull: "rpc error: code = Unknown desc = failed to pull and unpack image \"my.registry.tld/some/image:latest\": failed to resolve reference \"my.registry.tld/some/image:latest\": failed to authorize: failed to fetch oauth token: unexpected status: 401 Unauthorized"
```

**Additional context / logs:**

---

Any ideas how to debug this further?

**Comment by brandond** [comment]
Did you place registries.yaml on both nodes, and restart k3s? Can you show an example of pulling the image successfully with crictl?

**Comment by MyIgel** [comment]
Yes, it is on both nodes (and the service restarted).

And sure:
```
# crictl pull my.registry.tld/some/image:latest
Image is up to date for sha256:808aa00374cfcf2a4987a2299919df3edb3e7befb13b48709eb6842a4af57929
```

**Comment by chadzhao** [comment]
Try add secret by command : kubectl create secret ... and use imagePullSecrets.

It works for me.

**Comment by MyIgel** [comment]
Thats not an option here (due to changing environments) but having a dedicated "deploy" user is (which works using crictl) so i would suspect some k3s validation to have some problems here

**Comment by chadzhao** [comment]
I have no idea the difference with crictl, but I can tell what I find last night (when I try k3s first time):

1. First, I create secret by yaml, then use it via imagePullSecrets options, not working
``` yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-basic-auth
type: kubernetes.io/basic-auth
stringData:
  username: admin
  password: t0p-Secret
```
2. I tried /etc/rancher/k3s/registries.yaml, not working
3. I followed others suggestions, remove quote and slash in registries.yaml, no matter remove imagePullSecrets options or not, still not working 
``` yaml
configs:
  my.registry.tld:
    auth:
      username: user
      password: token
```
4. I create secret via command, this time it works.
```shell
kubectl create secret docker-registry secret-docker 
  --docker-username=user 
  --docker-password=pwd 
  --docker-email=a@abc.com
```

It may be helpful.

**Comment by EvanCarroll** [comment]
I was having this same problem too. In my case it was caused by duplicate default routes.

https://devops.stackexchange.com/a/16244/18965

**Comment by belugame** [comment]
I face this too now, and have also tried using `imagePullSecrets`. But can't get either way to work, while `ctr image pull` works fine.

```
ctr images pull -u gitlab-registry-token:mysecret git.myhost.com:5050/my-project/my-project/nginx:latest
git.myhost.com:5050/my-project/my-project	saved	
└──manifest (c38646cee002)              	already exists	
   └──config (fe8c257edc76)             	already exists	
application/vnd.docker.distribution.manifest.v2+json sha256:aaa646cee002ef4aaaac803f6f572152e7bacba0fa13da9e90121cc7d87ba
Pulling from OCI Registry (git.myhost.com:5050/my-project/my-project/nginx:latest)	elapsed: 1.3 s	total:  11.9 K	(8.9 KiB/s)	
```

My registries.yaml

```
cat /etc/rancher/k3s/registries.yaml 
mirrors:
  "git.myhost.com:5050":
     endpoint:
       - "https://git.myhost.com:5050"

configs:
  "git.myhost.com:5050":
     auth:
       username: "gitlab-registry-token"
       password: "mysecret"

```
In the events of the pod I see:

```
  Normal   Pulling    31m (x4 over 33m)    kubelet            Pulling image "git.myhost.com:5050/my-project/my-project/nginx:latest"
  Warning  Failed     31m (x4 over 33m)    kubelet            Failed to pull image "git.myhost.com:5050/my-project/my-project/nginx:latest": 
failed to pull and unpack image "git.myhost.com:5050/my-project/my-project/nginx:latest": failed to resolve reference "git.myhost.com:5050/my-project/my-project/nginx:latest": 
failed to authorize: failed to fetch anonymous token: unexpected status from GET request to
https://git.myhost.com/jwt/auth?scope=repository%3Amy-project%2Fmy-project%2Fnginx%3Apull&service=container_registry: 403 Forbidden    
```    

I restarted the k3s service after editing registries.yaml and I can see that it puts the credentials into /var/lib/rancher/k3s/agent/etc/containerd/config.toml

I don't see what I am missing or where I am going wrong. I can confirm that other containers can reach my git server.

**Comment by belugame** [comment]
This generated file looks kind of wrong:

/var/lib/rancher/k3s/agent/etc/containerd/certs.d/git.myhost/hosts.toml 

```
# File generated by k3s. DO NOT EDIT.
server = "https://git.myhost.com:5050/v2"
capabilities = ["pull", "resolve", "push"]

[host]
```
That is the entire content

**Comment by brandond** [comment]
That is the correct content given your configuration.

Your `mirrors` entry for your repo is pointless - the only endpoint you've listed is the default endpoint, so it is ignored.

Are you using a custom containerd config file? What do see in `/var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl` and `/var/lib/rancher/k3s/agent/etc/containerd/config.toml`?

**Comment by belugame** [comment]
Thank you for looking into it, @brandond 
**It started working now, but I have no clue why**, because I did not touch the auth, network or image parts at all when I retried pulling the 100ths time. Worried that it will come back :|

It is true that the JWT url `https://git.myhost.com/jwt/auth?scope=repository%3Amy-project%2Fmy-project%2Fnginx%3Apull&service=container_registry: 403 Forbidden` does not hold the port. Since I don't define that url anywhere I assumed JWT and registry are simply not on the same port for Gitlab.

The mirror part I only added out of desperation to be honest, I removed it again. 

I did create a config.toml.tmpl after this post here to increase the logging level.

```
cat /var/lib/rancher/k3s/agent/etc/containerd/config.toml.tmpl
{{ template "base" . }}
[debug]
  level = "debug"
```

```
cat /var/lib/rancher/k3s/agent/etc/containerd/config.toml
# File generated by k3s. DO NOT EDIT. Use config.toml.tmpl instead.
version = 2
root = "/var/lib/rancher/k3s/agent/containerd"
state = "/run/k3s/containerd"

[grpc]
  address = "/run/k3s/containerd/containerd.sock"

[plugins."io.containerd.internal.v1.opt"]
  path = "/var/lib/rancher/k3s/agent/containerd"

[plugins."io.containerd.grpc.v1.cri"]
  stream_server_address = "127.0.0.1"
  stream_server_port = "10010"
  enable_selinux = false
  enable_unprivileged_ports = true
  enable_unprivileged_icmp = true
  device_ownership_from_security_context = false
  sandbox_image = "rancher/mirrored-pause:3.6"

[plugins."io.containerd.grpc.v1.cri".containerd]
  snapshotter = "overlayfs"
  disable_snapshot_annotations = true

[plugins."io.containerd.grpc.v1.cri".cni]
  bin_dir = "/var/lib/rancher/k3s/data/cni"
  conf_dir = "/var/lib/rancher/k3s/agent/etc/cni/net.d"

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
  runtime_type = "io.containerd.runc.v2"

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
  SystemdCgroup = true

[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runhcs-wcow-process]
  runtime_type = "io.containerd.runhcs.v1"

[plugins."io.containerd.grpc.v1.cri".registry]
  config_path = "/var/lib/rancher/k3s/agent/etc/containerd/certs.d"

[plugins."io.containerd.grpc.v1.cri".registry.configs."git.smart1-dev.eu:5050".auth]
  username = "gitlab-registry-token"
  password = "mysecret"

[debug]
  level = "debug"
```

**Comment by belugame** [comment]
The issue/error message came back for me after adding additional nodes to my k3s cluster and it tried to pull images onto the new node. Copying the /etc/rancher/k3s/registries.yaml to the new node was not enough.
I manually edited /var/lib/rancher/k3s/agent/etc/containerd/config.toml
and added the lines I found in my primary node:

```
[plugins.'io.containerd.cri.v1.images'.registry.configs.'git.myhost.com:5050'.auth]
  username = "my-gitlab-token"
  password = "mysecrettoken"
```

and then after:

`sudo systemctl restart k3s-agent`

it worked for the new node.
